ini.cpp

```diff
@@ -0,0 +1,233 @@
+            config = config->next;
+            free(tmp_config);
+        }
+
+        free(section->name);
+        section = section->next;
+        free(tmp);
+    }
+}
+
+int IniFile::parse_file()
+{
+    FILE *file;
+    char *entry = NULL;
+    int line = 0, ret = 0;
+    size_t n = 0;
+    ssize_t read;
+
+    assert(filename);
+
+    file = fopen(filename, "r");
+    if (!file) {
+        log_error_errno(errno, "Could not open ini file '%s' (%m)", filename);
+        return -EIO;
+    }
+
+    while ((read = getline(&entry, &n, file)) >= 0) {
+        line++;
+
+        // Remove trailing '/n'
+        if (*(entry + read - 1) == '\n')
+            *(entry + read - 1) = '\0';
+
+        switch (entry[0]) {
+        case ';':
+        case '\0':
+            // Discards comment or blank line
+            free(entry);
+            break;
+        case '[':
+            ret = add_section(entry, line);
+            if (ret < 0) {
+                goto end;
+            }
+            break;
+        default:
+            ret = add_config(entry, line);
+            if (ret < 0) {
+                goto end;
+            }
+        }
+
+        entry = NULL;
+    }
+
+    current_section = nullptr;
+
+end:
+    free(entry);
+    fclose(file);
+    return ret;
+}
+
+int IniFile::add_section(char *entry, int line)
+{
+    struct section *section;
+
+    char *end = strchr(entry, ']');
+    if (!end) {
+        log_error("On file %s: Line %d: Unfinished section name. Expected ']'", filename, line);
+        return -EINVAL;
+    }
+
+    // So section name ends before ']'
+    *end = '\0';
+
+    // Ensure section is not duplicated. If it is,
+    // just make parser add on known section
+    section = sections;
+    while (section) {
+        if (!strcmp(section->name, entry)) {
+            current_section = section;
+            free(entry);
+            return 0;
+        }
+        section = section->next;
+    }
+
+    // Section is new.
+    section = (struct section *)calloc(1, sizeof(struct section));
+    null_check(section, -ENOMEM);
+
+    section->name = entry;
+    section->next = sections;
+    sections = section;
+
+    current_section = section;
+
+    return 0;
+}
+
+int IniFile::add_config(char *entry, int line)
+{
+    char *equal_pos;
+    struct config *config;
+
+    if (!(equal_pos = strchr(entry, '='))) {
+        log_error("On file %s: Line %d: Missing '=' on config", filename, line);
+        return -EINVAL;
+    }
+
+    if (equal_pos == entry) {
+        log_error("On file %s: Line %d: Missing name on config", filename, line);
+        return -EINVAL;
+    }
+
+    if (equal_pos == (entry + strlen(entry) - 1)) {
+        log_error("On file %s: Line %d: Missing value on config", filename, line);
+        return -EINVAL;
+    }
+
+    config = (struct config *)malloc(sizeof(struct config));
+    null_check(config, -ENOMEM);
+
+    config->key = entry;
+    config->value = equal_pos + 1;
+    *equal_pos = '\0';
```

Comment:
it's very common to have
Bla = "bli"

in this case you will have an extra space in the key.